spock-core/src/main/java/org/spockframework/runtime/extension/IStore.java

```diff
@@ -0,0 +1,323 @@
+   * for the store's {@code Namespace} unless they overwrite it.
+   *
+   * <p>If the {@code value} is an instance of {@link AutoCloseable}
+   * the {@code close()} method will be invoked on the stored object when
+   * the store is closed.
+   *
+   * <p>Any existing value in the current {@link IStore} will be replaced,
+   * byt not in ancestors. In addition, the {@link AutoCloseable} API will not
+   * be honored for values that are replaced via this method.
+   *
+   * @param key   the key under which the value should be stored; never
+   *              {@code null}
+   * @param value the value to store; may be {@code null}
+   * @return the previous value; may be {@code null}
+   * @see AutoCloseable
+   */
+  Object put(Object key, Object value);
+
+  /**
+   * Remove the value that was previously stored under the supplied {@code key}.
+   *
+   * <p>The value will only be removed in the current {@link IStore},
+   * not in ancestors. In addition, the {@link AutoCloseable} API will not
+   * be honored for values that are manually removed via this method.
+   *
+   * <p>For greater type safety, consider using {@link #remove(Object, Class)}
+   * instead.
+   *
+   * @param key the key; never {@code null}
+   * @return the previous value or {@code null} if no value was present
+   * for the specified key
+   * @see #remove(Object, Class)
+   */
+  Object remove(Object key);
+
+  /**
+   * Remove the value of the specified required type that was previously stored
+   * under the supplied {@code key}.
+   *
+   * <p>The value will only be removed in the current {@link IStore},
+   * not in ancestors. In addition, the {@link AutoCloseable} API will not
+   * be honored for values that are manually removed via this method.
+   *
+   * @param key          the key; never {@code null}
+   * @param requiredType the required type of the value; never {@code null}
+   *                     will throw {@link StoreException} if value isn't an instance of this type
+   * @param <V>          the value type
+   * @return the previous value or {@code null} if no value was present
+   * for the specified key
+   * @see #remove(Object)
+   */
+  <V> V remove(Object key, Class<V> requiredType);
+
+  /**
+   * A {@code Namespace} is used to provide a <em>scope</em> for data saved by
+   * extensions within an {@link IStore}.
+   *
+   * <p>Storing data in custom namespaces allows extensions to avoid accidentally
+   * mixing data between extensions or across different invocations within the
+   * lifecycle of a single extension.
+   */
+  class Namespace {
+
+    /**
+     * Create a namespace which restricts access to data to all extensions
+     * which use the same sequence of {@code parts} for creating a namespace.
+     *
+     * <p>The order of the {@code parts} is significant.
+     *
+     * <p>Internally the {@code parts} are compared using {@link Object#equals(Object)}.
+     */
+    public static Namespace create(Object... parts) {
```

Comment:
Shouldn't this document that the used Objects for parts, need to be `Immutable` or at least have a stable equals/hashcode contract.
If an object changes its hashcode I guess the thing is broken.
Or do we say, yeah if you do such things, that is not our problem...